Add amount customisation setting (fix #1235)

CPatchane opened the pull request:
Fix https://github.com/rotki/rotki/issues/1235

Add a new setting allowing amount customisation using a similar string with placeholders than for the date: %T for thousands, %U for units, %D for decimals and %C for the currency.
<img width="570" alt="Screenshot 2020-08-04 at 00 58 23" src="https://user-images.githubusercontent.com/10224453/89235221-146f6980-d5ee-11ea-8515-e458eb44698f.png">

This string system is cool to have one setting stored and handled but it might look complex on a UX point of view, I added a hint but it might not be enough. It would be great to discuss on that point imho.

- [x] Update and add new tests
- [x] Add some validations (front and backend)
- [x] Add a hint below the input to give a exemple with the current entered string (or invalid if not correct)
- [x] Replace to use it as the previous `formatPrice` filter

PS: I replaced `roundDown` filter thinking that it was not used anymore but I clearly may be wrong about that.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/rotki/rotki?pullRequest=1269) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/rotki/rotki?pullRequest=1269) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/rotki/rotki?pullRequest=1269) it.</sub>

CPatchane commented on frontend/app/src/components/taxreport/TaxReportEvents.vue:
```diff
@@ -18,27 +18,63 @@
             </template>
             <template #item.paidInProfitCurrency="{ item }">
-              {{ item.paidInProfitCurrency | formatPrice(floatingPrecision) }}
+              {{
+                item.paidInProfitCurrency
+                  | formatPrice(
+                    amountDisplayFormat,
+                    floatingPrecision,
+                    currency.unicode_symbol
+                  )
+              }}
```
Some DRY code here that we can clean by putting that on a component method?

kelsos replied:
It seems that we missed that when we introduced the `AmountDisplay.vue`, we can create a followup issue for replacing that with the component.

isidorosp replied:
Why don't we just introduce using AmountDisplay for these components with this PR? It would amount to the same thing.

kelsos replied:
If CPatchane wants to do that as part of his PR sure. But since this was not in the original scope of the issue, we can still merge and follow up with a new issue.

isidorosp replied:
He's already making changing to TaxReportEvents so that it works with this PR (which is outside the scope of this issue), it doesn't make sense to make changes that will just immediately be overwritten with another issue.

CPatchane replied:
I changed this component to use the `AmountDisplay.vue` component here

kelsos commented on frontend/app/src/data/amount_formatter.ts:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * Then we replace the currency placeholder
+     */
+    formattedString = formattedString.replace(/%C/gm, currency || '');
```
nitpick/question: any reason why not doing a return `format.replace(...).replace(...).trim()`

kelsos commented on frontend/app/src/views/settings/GeneralSettings.vue:
```diff
@@ -138,6 +138,29 @@
+              :messages="
+                'Resulting: ' +
+                (isValidAmountDisplayFormat(amountDisplayFormat)
+                  ? $options.filters.formatPrice(
```
Would it be possible to move this check/calculation in a computed property?

CPatchane replied:
Is this what you were thinking about? https://github.com/rotki/rotki/pull/1269/files#diff-fe784788b0a861bf65d37abcf3e8ec8dR239-R251

CPatchane replied:
It's now using directly the AmountDisplay.vue component here

isidorosp commented on frontend/app/src/components/display/AmountDisplay.vue:
```diff
@@ -13,17 +13,17 @@
       <span
         v-else-if="fiatCurrency === currency.ticker_symbol"
         class="amount-display__value"
       >
-        {{ renderValue | roundDown(floatingPrecision) }}
+        {{ renderValue | formatPrice(amountDisplayFormat, floatingPrecision) }}
       </span>
       <span v-else class="amount-display__value">
-        {{ renderValue | formatPrice(floatingPrecision) }}
+        {{ renderValue | formatPrice(amountDisplayFormat, floatingPrecision) }}
```
This the `else-if` and `else` clauses here now do the same thing. Previously this (and the original v-if) determined whether we were rounding up/down.
```
      <span
        v-else-if="fiatCurrency === currency.ticker_symbol"
        class="amount-display__value"
      >
        {{ renderValue | formatPrice(amountDisplayFormat, floatingPrecision) }}
      </span>
      <span v-else class="amount-display__value">
        {{ renderValue | formatPrice(amountDisplayFormat, floatingPrecision) }}
      </span>
```

kelsos replied:
I would guess we will still want the old behaviour when it comes to rounding, since we already had lengthy discussions on that and decided on this approach.

Also for the currency display, I agree it would make sense to have the handling as part of the amount display.

CPatchane replied:
I wasn't aware of this specific behaviour, so I have to get back that difference using formatPrice 👍

kelsos commented:
> @CPatchane This is some great work, thank you!
> 
> I have a few notes:
> 
>     * I recall that for cryptocurrencies we had agreed to round up (and show asterisk) but for fiat we were rounding down (no asterisk). With this change we are rounding down for everything (and for crypto amounts showing asterisk). Are we aligned on this? @LefterisJP @kelsos
>       ![image](https://user-images.githubusercontent.com/27592/89266394-283ebe00-d636-11ea-9387-94d892d51ba7.png)
> 
>     * What is the "use example" of this via AmountDisplay? From what I can see in the AmountDisplay code we never call `formatPrice()` with a currency argument provided. The "original" way we were applying currency is still in AmountDisplay (see below) -- i.e. if you specify `show-currency` or `asset` it will still put the currency/asset at the end. If the idea is to incorporate the name of the currency/asset into the format of the value, shouldn't we be replacing this portion of the component?
> 
> 
> NOTE: For what it's worth I would prefer if we somehow keep the "rendering" of asset/currency separate to the value in a separate `<span>` which would allow for (a) being able to grab the value and only the value for tests and (b) separate styling if needed. This was my main objection to storing the entire format as a "string", because then you would have to serialize/deserialize the position of the currency/amount when figuring out how to render the full component.
> 
> ```
>       <span
>         v-if="showCurrency === 'ticker' && !renderValue.isNaN()"
>         class="amount-display__currency"
>       >
>         {{ currency.ticker_symbol }}
>       </span>
>       <span
>         v-else-if="showCurrency === 'symbol' && !renderValue.isNaN()"
>         class="amount-display__currency"
>       >
>         {{ currency.unicode_symbol }}
>       </span>
>       <span
>         v-else-if="showCurrency === 'name' && !renderValue.isNaN()"
>         class="amount-display__currency"
>       >
>         {{ currency.name }}
>       </span>
>       <span
>         v-else-if="!!asset && !renderValue.isNaN()"
>         class="amount-display__asset"
>       >
>         {{ asset }}
>       </span>
> ```

I would guess we will still want the old behaviour when it comes to rounding, since we already had lengthy discussions on that and decided on this approach.

Also for the currency display, I agree it would make sense to have the handling as part of the amount display.

LefterisJP commented on rotkehlchen/api/v1/encoding.py:
```diff
@@ -698,6 +698,14 @@ class ModifiableSettingsSchema(Schema):
+    amount_display_format = fields.String(
+        strict=True,
+        validate=webargs.validate.Regexp(
+            regex=r".*%T.*%U.*%D.*",
```
So the currency format placeholder is optional?

CPatchane replied:
I guess it could be optional. But I may be wrong, do we want it mandatory?

LefterisJP replied:
no it's fine I guess. Defaults to being at the end I suppose.

LefterisJP commented:
Also @CPatchane don't forget to sign the CLA so that the PR can be merged when it's ready. Its purpose is to allow us to be able to change the license in the future if it's needed without having to ask each and every contributor.

LefterisJP commented:
Due to the amount of changes and things that could be affected we should merge this **after** v1.6.2 is released.

CPatchane commented:
I added a new commit which revamp the amount format setting to split to 3 settings: `thousandSeparator`, `decimalSeparator` and `currencyLocation`

Also, I changed the `AmountDisplay.vue` to handle correctly the currency location.

Finally, I build a specific card in the settings view to be dedicated to amounts settings:
<img width="615" alt="Screenshot 2020-08-09 at 02 18 20" src="https://user-images.githubusercontent.com/10224453/89722173-a14b6600-d9e6-11ea-9051-92031cb95d3e.png">

kelsos commented on frontend/app/src/components/taxreport/TaxReportOverview.vue:
```diff
@@ -15,11 +15,7 @@
-                      | calculatePrice(exchangeRate(currency.ticker_symbol))
-                      | formatPrice(floatingPrecision)
-                  }}
+                  <amount-display :amount="item"></amount-display>
```
This will only display the value but it, won't do the price conversion calculation right? it also needs the  `fiat-currency="USD"` so that it converts the value to the selected currency.

kelsos replied:
Give me a moment to clarify that though before doing anything

kelsos replied:
Ignore that comment apparently that was a bug.

kelsos commented on frontend/app/src/components/taxreport/TaxReportEvents.vue:
```diff
@@ -18,28 +18,28 @@
             <template #item.paidInProfitCurrency="{ item }">
-              {{ item.paidInProfitCurrency | formatPrice(floatingPrecision) }}
+              <amount-display
+                :amount="item.paidInProfitCurrency"
```
Also for this, in currency means that you also need to add `fiat-currency="USD"` to have the proper exchange rate conversion happen.

kelsos commented on frontend/app/src/components/display/AmountCurrency.vue:
```diff
@@ -0,0 +1,50 @@
+<template>
+  <fragment>
```
Did you get errors about a single root element in this case? I remember having no issues with `v-if`. Still, this is useful until the vue3 update, we have a lot of cases of divs that are used only due to the single root element limitation.

CPatchane replied:
Oh I didn't try, I thought that would be an issue, good to know then, do you prefer I change for here?

kelsos replied:
No need, just leave it be, we can use it in other places as well :)

LefterisJP commented:
Ze merging will happen
